httpbp/client_middlewares.go

```diff
@@ -0,0 +1,222 @@
+	return transport
+}
+
+// ClientErrorWrapper applies ClientErrorFromResponse to the returned response
+// ensuring a HTTP status response outside the range [200, 400) is wrapped in
+// an error relieving users from the need to check the status response.
+//
+// Adding this middleware means that both, non-nil response and non-nil error,
+// can be returned. If configured the caller needs to call
+// httpbp.DrainAndClose(resp.Body) to ensure the underlying TCP connection can
+// be re-used.
+func ClientErrorWrapper() ClientMiddleware {
+	const limit = 1024
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripperFunc(func(req *http.Request) (resp *http.Response, err error) {
+			resp, err = next.RoundTrip(req)
+			if err != nil {
+				return resp, err
+			}
+			if err == nil {
+				err = ClientErrorFromResponse(resp)
+				if err != nil {
+					defer DrainAndClose(resp.Body)
+					var ce *ClientError
+					if !errors.As(err, &ce) {
+						return resp, err
+					}
+					body, e := io.ReadAll(io.LimitReader(resp.Body, limit))
+					if e != nil {
+						return resp, e
+					}
+					ce.AdditionalInfo = string(body)
+					return resp, ce
+				}
+			}
+			return resp, err
+		})
+	}
+}
+
+// CircuitBreaker is a middleware that prevents sending requests that are likely
+// to fail based on a configurable failure ratio based on total failures and
+// requests.
+func CircuitBreaker(config breakerbp.Config) ClientMiddleware {
+	pool := &sync.Pool{
+		New: func() interface{} {
+			return breakerbp.NewFailureRatioBreaker(config)
+		},
+	}
+	breakers := &sync.Map{}
+
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripperFunc(func(req *http.Request) (*http.Response, error) {
+			host := req.URL.Hostname()
+
+			// new circuit breakers should rarely get allocated: the pool reduces
+			// garbage collector overhead
+			newBreaker := pool.Get()
+			breaker, loaded := breakers.LoadOrStore(host, newBreaker)
+			if loaded {
+				defer pool.Put(newBreaker)
+			}
+
+			resp, err := breaker.(breakerbp.FailureRatioBreaker).Execute(func() (interface{}, error) {
+				resp, err := next.RoundTrip(req)
+				if err != nil {
+					return nil, err
+				}
+				// circuit break on any HTTP 5xx code
+				if resp.StatusCode >= http.StatusInternalServerError {
+					// read & close to ensure underlying RoundTripper
+					// (http.Transport) is able to re-use the persistent TCP
+					// connection.
+					_, _ = io.ReadAll(resp.Body)
+					return nil, resp.Body.Close()
+				}
+				return resp, nil
+			})
+			if err != nil {
+				return nil, err
+			}
+			return resp.(*http.Response), nil
+		})
+	}
+}
+
+// Retries provides a retry middleware.
+func Retries(retryOptions ...retry.Option) ClientMiddleware {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripperFunc(func(req *http.Request) (resp *http.Response, err error) {
+			if len(retryOptions) == 0 {
+				retryOptions = []retry.Option{retry.Attempts(1)}
+			}
```

Comment:
Another thing about retry http requests I just realized this morning: you can't really retry requests with a body (e.g. you can only retry get requests).

The reason is that [`http.Request.Body`](https://pkg.go.dev/net/http#Request.Body) is an `io.ReadCloser`, so you can only read it once. On retries, the read will return EOF immediately, and as a result, those retried requests will *silently* be corrupt (using empty body instead of the original body intended).

I think we should document that, *and* force no retry inside this middleware if `req.Body` is non-nil (we probably want to do some tests to make sure that `req.Body != nil` is a reliable way to distinguish the situation).

Reply:
actually never mind. I think that case is already handled because when redirect happens you still need to read the request body multiple times, so there's [`http.Request.GetBody`](https://pkg.go.dev/net/http#Request.GetBody).

we probably want to double check that this works, though.

Reply:
Added a test to check if it reads the body twice in case of a POST request.